pkg/kv/kvserver/loqrecovery/storage.go

```diff
@@ -0,0 +1,187 @@
+				filepath.Join(s.path, name), filepath.Join(s.path, info.Name()))
+		}
+		name = info.Name()
+		created = info.ModTime()
+		return nil
+	}); err != nil {
+		return StoredPlan{}, err
+	}
+	if len(name) > 0 {
+		return StoredPlan{
+			Name:         s.planFullName(name),
+			CreationTime: created,
+		}, nil
+	}
+	return StoredPlan{}, nil
+}
+
+func (s PlanStorage) RemovePlans() error {
+	return s.visitPlanFiles(func(info fs.FileInfo) error {
+		return s.fs.Remove(filepath.Join(s.path, info.Name()))
+	})
+}
+
+func (s PlanStorage) SavePlan(plan loqrecoverypb.ReplicaUpdatePlan) error {
+	existing, err := s.HasPlan()
+	if err != nil {
+		return errors.Wrap(err, "failed to perform existing recovery plan check")
+	}
+	if !existing.Empty() {
+		return errors.Newf("can not set recovery plan %s, plan file %s is already set", plan.PlanID,
+			existing.Name)
+	}
+
+	if err := s.ensureStoreExists(); err != nil {
+		return err
+	}
+
+	planBaseName := s.planFullName(plan.PlanID.String())
+	tmpFileName := planBaseName + ".tmp"
+	defer func() { _ = s.fs.Remove(tmpFileName) }()
+
+	if err := func() error {
+		outFile, err := s.fs.Create(tmpFileName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create file %q", tmpFileName)
+		}
+		defer func() { _ = outFile.Close() }()
+		jsonpb := protoutil.JSONPb{Indent: "  "}
+		var out []byte
+		if out, err = jsonpb.Marshal(plan); err != nil {
+			return errors.Wrap(err, "failed to marshal recovery plan")
+		}
+		if _, err = outFile.Write(out); err != nil {
+			return errors.Wrap(err, "failed to write recovery plan")
+		}
+		return nil
+	}(); err != nil {
+		return err
+	}
+
+	if err := s.fs.Rename(tmpFileName, planBaseName+".json"); err != nil {
+		_ = s.fs.Remove(tmpFileName)
+		return errors.Wrap(err, "failed to rename temp plan file")
+	}
+	return nil
+}
+
+func (s PlanStorage) LoadPlan(plan StoredPlan) (loqrecoverypb.ReplicaUpdatePlan, error) {
+	f, err := s.fs.Open(plan.Name)
+	if err != nil {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Wrapf(err, "failed to open plan file %q",
+			plan.Name)
+	}
+	defer func() { _ = f.Close() }()
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Wrapf(err, "failed to read plan file %q",
+			plan.Name)
+	}
+
+	var nodeUpdates loqrecoverypb.ReplicaUpdatePlan
+	jsonpb := protoutil.JSONPb{Indent: "  "}
+	if err = jsonpb.Unmarshal(data, &nodeUpdates); err != nil {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Wrapf(err,
+			"failed to unmarshal plan from file %q", plan.Name)
+	}
+
+	if s.planFullName(nodeUpdates.PlanID.String()+".json") != plan.Name {
+		return loqrecoverypb.ReplicaUpdatePlan{}, errors.Newf("loaded plan has id %s while filename is %s, plan id must match filename without extension",
+			nodeUpdates.PlanID, plan.Name)
+	}
+
+	return nodeUpdates, nil
+}
+
+func (s PlanStorage) visitPlanFiles(visitor func(fs.FileInfo) error) error {
```

Comment:
nit: Why the visitor pattern rather than just returning the list? We have to load the entire list into memory anyway.

Reply:
Dropped visitor to make it a simple loop over matching files.